### ISSUE TITLE
dynamic token

### ISSUE THREAD

**Issue by salvavidal** [adrienverge/openfortivpn#1142]
is there any solution to connect to fortigate if the system ask for token sent to your email when you try to connect?

**Comment by mrbaseman** [comment]
I think this is one variant of the OTP feature. Even in case the fortigate otp prompt is somehow non-standard, you can supply a string to search for in the data sent by the fortigate. Once openfortivpn finds this string, it asks you to enter the code and sends it to the fortigate. The default way is to install a fortitoken app or buy hardware tokens, but depending on the authentication infrastructure already in place, there may be other variants.

**Comment by Anat0l** [comment]
Let me rephrase the author's question. Is it possible to somehow transfer the PIN/OPT to a running program without restarting? After authorization, you need to enter a PIN/OPT, which is sent to your phone via SMS.

**Comment by DimitriPapadopoulos** [comment]
Let me rephrase the answer to the author's question. Yes, it should be possible, but you may have to supply a specific string for openfortivpn to look for and trigger the OTP mechanism.

It all happens in [auth_log_in()](https://github.com/adrienverge/openfortivpn/blob/master/src/http.c#L630) and the functions it calls:
https://github.com/adrienverge/openfortivpn/blob/2f9ae3d73e89fadf633b46e699a06fd396f40a97/src/http.c#L675-L682

By the way, see https://github.com/adrienverge/openfortivpn/wiki#reporting-issues.

**Comment by Anat0l** [comment]
Apparently you didn't quite understand. If you start the service with the -otp=123456 parameter at the time of startup, the token is recognized and sent at the right time, but it is not known in advance. If the token is not specified at the time of launch, then an empty string is sent to the token request coins and authorization fails.

Sorry, translated via Google

**Comment by DimitriPapadopoulos** [comment]
Apparently you didn't quite understand. Just don't start openfortivpn with the -otp=123456 parameter.

**Comment by Anat0l** [comment]
how to launch it so that you can enter or pass the opt parameter to the program after authorization?

**Comment by DimitriPapadopoulos** [comment]
Just run `openfortivpn`.

**Comment by salvavidal** [comment]
The issue is, when i run forticlient (windows interface) they send by mail the token which i have to input to start connection.

How can i do the same with openfortivpn ? using CLI how can i input this token? if they will send it after i send my connection request i don't find how to do it with openfortivpn.

Could i create a Script which run openfortivpn and then look forfor in my email the token but, could i be able to input it on anyplace?

i don't see clear.

**Comment by DimitriPapadopoulos** [comment]
Right, I cannot recall whether openfortivpn sends the password to the VPN server **before** asking for the OTP, or if it waits to have both before sending them to the server. If the latter is true, you will need to modify the source code accordingly.

**Comment by Anat0l** [comment]
> The issue is, when i run forticlient (windows interface) they send by mail the token which i have to input to start connection.
> 
> How can i do the same with openfortivpn ? using CLI how can i input this token? if they will send it after i send my connection request i don't find how to do it with openfortivpn.
> 
> Could i create a Script which run openfortivpn and then look forfor in my email the token but, could i be able to input it on anyplace?
> 
> i don't see clear.

I still couldn't solve this problem in a standard way. As the author wrote above, it is necessary to change the source code at the authorization stage.



> Right, I cannot recall whether openfortivpn sends the password to the VPN server **before** asking for the OTP, or if it waits to have both before sending them to the server. If the latter is true, you will need to modify the source code accordingly.

Yes, the code comes only after successful authorization using login and password.
